android/quest/src/main/java/org/smartregister/fhircore/quest/ui/launcher/GeoWidgetLauncherViewModel.kt

```diff
@@ -58,56 +55,71 @@ constructor(
     val defaultRepository: DefaultRepository,
     val dispatcherProvider: DispatcherProvider,
     val sharedPreferencesHelper: SharedPreferencesHelper,
+    val resourceDataRulesExecutor: ResourceDataRulesExecutor,
 ) : ViewModel() {
 
     private val _snackBarStateFlow = MutableSharedFlow<SnackBarMessageConfig>()
     val snackBarStateFlow = _snackBarStateFlow.asSharedFlow()
 
-    private val _locationsFlow: MutableStateFlow<Set<GeoWidgetLocation>> =
+    private val _locationsFlow: MutableStateFlow<Set<Feature>> =
         MutableStateFlow(setOf())
-    val locationsFlow: StateFlow<Set<GeoWidgetLocation>> = _locationsFlow
+    val locationsFlow: StateFlow<Set<Feature>> = _locationsFlow
 
     private val _locationDialog = MutableLiveData<String>()
     val locationDialog: LiveData<String> get() = _locationDialog
 
     // TODO: use List or Linkage resource to connect Location with Group/Patient/etc
-    private fun retrieveLocations() {
+    private fun retrieveLocations(geoWidgetConfig: GeoWidgetConfiguration) {
         viewModelScope.launch(dispatcherProvider.io()) {
-            val totalResource = defaultRepository.count(Search(ResourceType.Location))
-
-            val totalIteration = ceil(totalResource / PAGE_SIZE.toDouble()).toInt()
-
-            repeat(totalIteration) { index ->
-                val startingIndex = index * PAGE_SIZE
-                val search = Search(ResourceType.Location, PAGE_SIZE, startingIndex)
-
-              defaultRepository.search<Location>(search).forEach { location ->
-                  if (location.hasPosition() && location.position.hasLatitude() && location.position.hasLongitude()) {
-                      val geoWidgetLocation = GeoWidgetLocation(
-                          id = location.id,
-                          name = location.name ?: "",
-                          position = Position(
-                              location.position.latitude.toDouble(),
-                              location.position.longitude.toDouble()
-                          ),
-                          // TODO: add logic to decide the color of location
-                      )
-                      addLocationToFlow(geoWidgetLocation)
-                  }
-              }
-          }
-      }
-  }
-
-    fun checkSelectedLocation() {
+            val repositoryResourceDataList =
+                defaultRepository
+                    .searchResourcesRecursively(
+                        filterActiveResources = null,
+                        fhirResourceConfig = geoWidgetConfig.resourceConfig,
+                        configRules = null,
+                        secondaryResourceConfigs = null,
+                    )
```

Comment:
Loading all the Location data with the related resources may impact performance. Please highlight in a TODO for followup.